File: theory/src/main/kotlin/app/guitar/theory/Inversions.kt
package app.guitar.theory

/**
 * Chord-inversion helpers for the ear-training "Inversions" trainer.
 *
 * An inversion stacks the same chord tones but puts a different tone in the bass:
 *   - inversion 0 = root position (root in bass)
 *   - inversion 1 = 1st inversion (3rd in bass)
 *   - inversion 2 = 2nd inversion (5th in bass)
 *   - inversion 3 = 3rd inversion (7th in bass, 7th chords only)
 */
object Inversions {

    /** Number of distinct inversions a [quality] has (= its tone count). */
    fun count(quality: ChordQuality): Int = quality.intervals.size

    /** Human label for inversion index [k]. */
    fun name(k: Int): String = when (k) {
        0 -> "Root position"
        1 -> "1st inversion"
        2 -> "2nd inversion"
        3 -> "3rd inversion"
        else -> "${k}th inversion"
    }

    /**
     * MIDI notes for [quality] rooted at [rootMidi], voiced in [inversion]. The
     * bottom [inversion] chord tones are lifted an octave so the desired tone sits
     * in the bass; the result is returned low→high.
     */
    fun midis(rootMidi: Int, quality: ChordQuality, inversion: Int): List<Int> {
        val notes = quality.intervals.map { rootMidi + it.semitones }.toMutableList()
        val k = inversion.coerceIn(0, notes.size - 1)
        repeat(k) {
            val low = notes.removeAt(0)
            notes.add(low + 12)
        }
        return notes.sorted()
    }
}

File: chorect-web/src/theory/inversions.ts
// Chord-inversion helpers for the ear-training Inversions trainer, ported from
// theory/.../Inversions.kt.

import { ChordQuality } from "./chords";

export function inversionCount(quality: ChordQuality): number {
  return quality.intervals.length;
}

export function inversionName(k: number): string {
  switch (k) {
    case 0: return "Root position";
    case 1: return "1st inversion";
    case 2: return "2nd inversion";
    case 3: return "3rd inversion";
    default: return `${k}th inversion`;
  }
}

/** MIDI notes for [quality] rooted at [rootMidi], voiced in [inversion], low→high. */
export function inversionMidis(rootMidi: number, quality: ChordQuality, inversion: number): number[] {
  const notes = quality.intervals.map((iv) => rootMidi + iv);
  const k = Math.min(Math.max(inversion, 0), notes.length - 1);
  for (let i = 0; i < k; i++) {
    const low = notes.shift()!;
    notes.push(low + 12);
  }
  return notes.slice().sort((a, b) => a - b);
}
